plugins/kafka/kafka.cpp

```diff
@@ -0,0 +1,1167 @@
+        :   Thread("Kafka::Poller"),
+            parentPtr(_parentPtr),
+            pollTimeout(_pollTimeout)
+    {
+        shouldRun = false;
+        running = false;
+    }
+
+    Poller::~Poller()
+    {
+
+    }
+
+    void Poller::start()
+    {
+        if (!running)
+        {
+            shouldRun = true;
+            Thread::start();
+        }
+    }
+
+    void Poller::stop()
+    {
+        if (running)
+        {
+            shouldRun = false;
+
+            while (running)
+            {
+                usleep(pollTimeout);
+            }
+        }
+    }
+
+    int Poller::run()
+    {
+        if (parentPtr)
+        {
+            RdKafka::Handle*    handle = parentPtr->handle();
+
+            running = true;
+
+            while (shouldRun)
+            {
+                handle->poll(pollTimeout);
+            }
+        }
+
+        running = false;
+
+        return 0;
+    }
+
+    //--------------------------------------------------------------------------
+
+    Publisher::Publisher(const std::string& _brokers, const std::string& _topic, __int32 _pollTimeout)
+        :   brokers(_brokers),
+            topic(_topic),
+            pollTimeout(_pollTimeout)
+    {
+        producerPtr = NULL;
+        topicPtr = NULL;
+        lockPtr = new Mutex;
+        pollerPtr = new Poller(this, _pollTimeout);
+
+        updateTimeTouched();
+    }
+
+    Publisher::~Publisher()
+    {
+        if (pollerPtr)
+        {
+            delete(pollerPtr);
+            pollerPtr = NULL;
+        }
+
+        if (topicPtr)
+        {
+            delete(topicPtr);
+            topicPtr = NULL;
+        }
+
+        if (producerPtr)
+        {
+            delete(producerPtr);
+            producerPtr = NULL;
+        }
+
+        if (lockPtr)
+        {
+            delete(lockPtr);
+        }
+    }
+
+    RdKafka::Handle* Publisher::handle()
+    {
+        return static_cast<RdKafka::Handle*>(producerPtr);
+    }
+
+    time_t Publisher::updateTimeTouched()
+    {
+        timeCreated = time(NULL);
+
+        return timeCreated;
+    }
+
+    time_t Publisher::getTimeTouched() const
+    {
+        return timeCreated;
+    }
+
+    void Publisher::shutdownPoller()
+    {
+        if (pollerPtr)
+        {
+            // Wait until we send all messages
+            while (messagesWaitingInQueue() > 0)
+            {
+                usleep(pollTimeout);
+            }
+
+            // Tell poller to stop
+            pollerPtr->stop();
+        }
+    }
+
+    __int32 Publisher::messagesWaitingInQueue()
+    {
+        __int32 queueLength = 0;
+
+        if (producerPtr)
+        {
+            queueLength = producerPtr->outq_len();
+        }
+
+        return queueLength;
+    }
+
+    void Publisher::ensureSetup()
+    {
+        if (!topicPtr)
+        {
+            // Use a mutex to guarantee that multiple threads don't try to populate
+            // this object at the same time
+            lockPtr->lock();
```

Comment:
If you had used a CriticalSection rather than a Mutex, you could have used CriticalBlock here and not needed to worry about making sure it was released on exceptions etc...